File: java/src/main/java/ch/frostnova/example/shoppingcart/util/Check.java
package ch.frostnova.example.shoppingcart.util;

import java.math.BigDecimal;
import java.util.regex.Pattern;

public final class Check {

    private Check() {

    }

    @SafeVarargs
    public static <T> T required(String property, T value, Validator<? super T>... validators) {
        if (value == null) {
            throw new ValidationException(property, "is required");
        }
        if (value instanceof String && ((String) value).length() == 0) {
            throw new ValidationException(property, value, "must not be empty");
        }
        for (Validator<? super T> validator : validators) {
            String error = validator.validate(value);
            if (error != null) {
                throw new ValidationException(property, value, error);
            }
        }
        return value;
    }

    public static Validator<Number> min(Number min) {
        return value -> new BigDecimal(value.toString()).compareTo(new BigDecimal(min.toString())) < 0
                ? String.format("must be at least %s", min) : null;
    }

    public static Validator<String> format(String regex) {
        return format(regex, "invalid format");
    }

    public static Validator<String> format(String regex, String error) {
        Pattern pattern = Pattern.compile(regex);
        return value -> !pattern.matcher(value).matches() ? error : null;
    }

    @FunctionalInterface
    public interface Validator<T> {
        String validate(T value);
    }

    public static final class ValidationException extends RuntimeException {

        public ValidationException(String property, String error) {
            super(String.format("%s %s", property, error));
        }

        public ValidationException(String property, Object value, String error) {
            super(String.format("%s: '%s' %s", property, value, error));
        }
    }
}

File: kotlin/src/main/kotlin/ch/frostnova/example/shoppingcart/util/Check.kt
package ch.frostnova.example.shoppingcart.util

import java.math.BigDecimal
import java.util.regex.Pattern

object Check {

    @JvmStatic
    @SafeVarargs
    fun <T> required(property: String, value: T, vararg validators: Validator<in T>): T {
        if (value is String && (value as String).length == 0) {
            throw ValidationException(property, value, "must not be empty")
        }
        for (validator in validators) {
            val error = validator.validate(value)
            if (error != null) {
                throw ValidationException(property, value, error)
            }
        }
        return value
    }

    fun min(min: Number): Validator<Number> {
        return object : Validator<Number> {
            override fun validate(value: Number): String? {
                return if (BigDecimal(value.toString()) < BigDecimal(min.toString()))
                    "must be at least $min" else null
            }
        }
    }

    @JvmOverloads
    fun format(regex: String?, error: String = "invalid format"): Validator<String> {
        val pattern = Pattern.compile(regex)
        return object : Validator<String> {
            override fun validate(value: String): String? {
                return if (!pattern.matcher(value).matches()) error else null
            }
        }
    }

    @FunctionalInterface
    interface Validator<T> {
        fun validate(value: T): String?
    }

    class ValidationException(property: String?, value: Any?, error: String?) : RuntimeException("$property: '$value' $error")
}